csrc/loader.cpp

```diff
@@ -60,3 +64,47 @@ JNIEXPORT jstring JNICALL Java_com_amazon_corretto_crypto_provider_Loader_getNat
     }
 
 }
+
+JNIEXPORT jboolean JNICALL Java_com_amazon_corretto_crypto_provider_Loader_validateLibcryptoExactVersionMatch(JNIEnv* pEnv, jclass)
+{
+    char msg_buffer[256] = {0};
+
+    try {
+        const unsigned long libcrypto_compiletime_version = OPENSSL_VERSION_NUMBER;
+        const unsigned long libcrypto_runtime_version = OpenSSL_version_num();
+
+        if (libcrypto_compiletime_version != libcrypto_runtime_version) {
+            snprintf(msg_buffer, sizeof(msg_buffer), "Runtime libcrypto version does not match compile-time version. "
+                "Expected: 0x%08lX , Actual: 0x%08lX", libcrypto_compiletime_version, libcrypto_runtime_version);
+            throw java_ex(EX_RUNTIME_CRYPTO, msg_buffer);
+        }
+
+        return JNI_TRUE;
+    } catch (java_ex &ex) {
+        ex.throw_to_java(pEnv);
+    }
+
+    return JNI_FALSE;
+}
+
+JNIEXPORT jboolean JNICALL Java_com_amazon_corretto_crypto_provider_Loader_validateLibcryptoFuzzyVersionMatch(JNIEnv* pEnv, jclass)
```

Comment:
minor style nit: decent amount of code duplicated between fuzzy/exact matchers. you could pull out most of each of these functions into a shared helper that takes compile-time and runtime versions and returns a boolean.